server/master/src/main/java/org/apache/accumulo/master/upgrade/UpgradeCoordinator.java

```diff
@@ -22,47 +22,120 @@
+     * nothing is upgraded when seeing this.
+     */
+    INITIAL {
+      @Override
+      public boolean isParentLevelUpgraded(KeyExtent extent) {
+        return false;
+      }
+    },
+    /**
+     * This signifies that only zookeeper has been upgraded so far.
+     */
+    UPGRADED_ZOOKEEPER {
+      @Override
+      public boolean isParentLevelUpgraded(KeyExtent extent) {
+        return extent.isRootTablet();
+      }
+    },
+    /**
+     * This signifies that only zookeeper and the root table have been upgraded so far.
+     */
+    UPGRADED_ROOT {
+      @Override
+      public boolean isParentLevelUpgraded(KeyExtent extent) {
+        return extent.isMeta();
+      }
+    },
+    /**
+     * This signifies that everything (zookeeper, root table, metadata table) is upgraded.
+     */
+    COMPLETE {
+      @Override
+      public boolean isParentLevelUpgraded(KeyExtent extent) {
+        return true;
+      }
+    },
+    /**
+     * This signifies a failure occurred during upgrade.
+     */
+    FAILED {
+      @Override
+      public boolean isParentLevelUpgraded(KeyExtent extent) {
+        return false;
+      }
+    };
+
+    /**
+     * Determines if the place where this extent stores its metadata was upgraded for a given
+     * upgrade status.
+     */
+    public abstract boolean isParentLevelUpgraded(KeyExtent extent);
+  }
+
   private static Logger log = LoggerFactory.getLogger(UpgradeCoordinator.class);
 
-  private ServerContext context;
-  private boolean haveUpgradedZooKeeper = false;
-  private boolean startedMetadataUpgrade = false;
   private int currentVersion;
   private Map<Integer,Upgrader> upgraders = Map.of(ServerConstants.SHORTEN_RFILE_KEYS,
       new Upgrader8to9(), ServerConstants.CRYPTO_CHANGES, new Upgrader9to10());
 
-  public UpgradeCoordinator(ServerContext ctx) {
-    int currentVersion = ServerUtil.getAccumuloPersistentVersion(ctx.getVolumeManager());
+  private volatile UpgradeStatus status;
 
-    ServerUtil.ensureDataVersionCompatible(currentVersion);
+  public UpgradeCoordinator() {
+    status = UpgradeStatus.INITIAL;
+  }
 
-    this.currentVersion = currentVersion;
-    this.context = ctx;
+  private void setStatus(UpgradeStatus status, EventCoordinator eventCoordinator) {
+    UpgradeStatus oldStatus = this.status;
+    this.status = status;
+    // calling this will wake up threads that may assign tablets. After the upgrade status changes
+    // those threads may make different assignment decisions.
+    eventCoordinator.event("Upgrade status changed from %s to %s", oldStatus, status);
   }
 
   @SuppressFBWarnings(value = "DM_EXIT",
       justification = "Want to immediately stop all master threads on upgrade error")
   private void handleFailure(Exception e) {
     log.error("FATAL: Error performing upgrade", e);
+    // do not want to call setStatus and signal an event in this case
+    status = UpgradeStatus.FAILED;
     System.exit(1);
```

Comment:
I wonder if we still need to exit here to force threads to stop, now that things are being upgraded more gracefully and we're not moving on to the next step until we complete the prior upgrade steps. Probably not... but that can be pursued apart from this PR.

Reply:
Things may be in a really bad state and who knows what other threads are doing.  I think this is best.  I would only advocate for removing it if we knew what every other thread in the process was doing when this happened.